### DISCUSSION TITLE
MagicByte exception when trying to render .osm file

### DISCUSSION THREAD

**Discussion by cesarmatt** [mapsforge/mapsforge#1224] [Q&A] [answered]
Hello!

I'm having issues with MagicByte while trying to render a .osm file with mapsforge.

Following this GettingStarted.java(https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/GettingStarted.java) was possible for me to render .map files but my request to backend should return a .osm file.

I tried downloading a .osm from Geofabrik and generating a custom file with Overpass but both ended up failing when trying to create MapFile as shown at line 115 of GettingStarted.java.

I'm wondering if there is a way to render a .osm file with mapsforge or if I'm stuck to .map.

Here is my code:
```
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        AndroidGraphicFactory.createInstance(applicationContext)

        val binding = ActivityInspectionBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupViews()
        setupBindings()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == SELECT_MAP_FILE && resultCode == Activity.RESULT_OK) {
            if (data != null) {
                val uri = data.data
                if (uri != null) {
                    setupMap(uri)
                }
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    private fun setupMap(uri: Uri) {
        try {
            mapView?.mapScaleBar?.isVisible = true
            mapView?.setBuiltInZoomControls(true)

            val tileCache = AndroidUtil.createTileCache(
                this,
                "mapcache",
                    mapView?.model?.displayModel?.tileSize!!,
                1f,
                    mapView?.model?.frameBufferModel?.overdrawFactor!!
                )

            val fileInputStream = contentResolver.openInputStream(uri) as FileInputStream
            val mapDataStore = MapFile(fileInputStream)
            val tileRendererLayer = TileRendererLayer(
                tileCache,
                mapDataStore,
                mapView?.model?.mapViewPosition,
                AndroidGraphicFactory.INSTANCE
            )

            tileRendererLayer.setXmlRenderTheme(InternalRenderTheme.OSMARENDER)

            mapView?.layerManager?.layers?.add(tileRendererLayer)

            mapView?.setCenter(LatLong(-34.929545, -56.158769))
            mapView?.setZoomLevel(15)
        } catch (exception: Exception) {
            exception.printStackTrace()
        }
    }
```
Error being logged at logcat:
`org.mapsforge.map.reader.header.MapFileException: invalid magic byte: ﻿<?xml version="1.`

Files I tried to render:
https://drive.google.com/file/d/1R6zFxaPMWVRk_0EdcTN5ibyNAgqtz6No/view?usp=sharing

Since I couldn't find many support materials online you guys are my last resort 😛 

Thanks for the help, have a nice day!

**Comment by devemux86** [comment, marked as answer]
_First you need to create a Mapsforge map file from the osm/pbf and then open the map file with the library._

There is a detailed [Map Creation Guide](https://github.com/mapsforge/mapsforge/blob/master/docs/MapCreation.md) mentioned in the main page.

It uses the map-writer tool, see its [documentation](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md).

Mapsforge maps can also be created automatically with [mapsforge-creator](https://github.com/mapsforge/mapsforge-creator).

**Reply by cesarmatt** [reply]
Thanks for the quick response!
There is any lib to render .osm files instead of Mapbox? I'm a bit unsure about Mapbox 'cause they're a paid solution.

**Reply by devemux86** [reply]
Map libraries can read OSM data after they are converted to a suitable map format.

You can check [VTM](https://github.com/mapsforge/vtm), an OpenGL map library (compatible with Mapsforge maps).
